### PR TITLE
V request context

### PR DESCRIPTION
**Pull request by sevikerr**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+##[Unreleased] 2016-05-25
+### Added
+- #25 class RequestContext (alternate of `setData` function in `Form` class)
```
Comment (funivan) [inline CHANGELOG.md]: #25 new class `RequestContext`

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,19 @@
...
+- #25 function `handleRequest($request)` to `Form` class
+- #25 function `handleRequest($request)` to `BaseElement` class
+- #25 function `clearValue()` to `ElementInterface`
+- function `isChecked($value)` to `CheckBoxList` class
+- more tests for `\Fiv\Form\Element\...` classes
+### Deprecated
+- #25 Form method `setData` (migration to using `RequestContext` class)
```
Comment (funivan) [inline CHANGELOG.md]: #25 Deprecate method `Form::setData`  use ...... Write information about new method

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+##[Unreleased] 2016-05-25
+### Added
+- #25 class RequestContext (alternate of `setData` function in `Form` class)
+- #25 function `handleRequest($request)` to `Form` class
+- #25 function `handleRequest($request)` to `BaseElement` class
+- #25 function `clearValue()` to `ElementInterface`
+- function `isChecked($value)` to `CheckBoxList` class
+- more tests for `\Fiv\Form\Element\...` classes
```
Comment (funivan) [inline CHANGELOG.md]: remove this line

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+##[Unreleased] 2016-05-25
+### Added
+- #25 class RequestContext (alternate of `setData` function in `Form` class)
+- #25 function `handleRequest($request)` to `Form` class
+- #25 function `handleRequest($request)` to `BaseElement` class
+- #25 function `clearValue()` to `ElementInterface`
+- function `isChecked($value)` to `CheckBoxList` class
```
Comment (funivan) [inline CHANGELOG.md]: ??

File: src/Element/BaseElement.php [hunk 5]
```diff
@@ -92,6 +93,17 @@ public function setValue($value) {
 
 
     /**
+     * @inheritdoc
+     */
+    public function clearValue() {
```
Comment (funivan) [inline src/Element/BaseElement.php]: we can remove this logic and always set value. But value can also be null

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,19 @@
 # Changelog
 
+##[Unreleased] 2016-05-25
+### Added
+- #25 class RequestContext (alternate of `setData` function in `Form` class)
+- #25 function `handleRequest($request)` to `Form` class
```
Comment (funivan) [inline CHANGELOG.md]: handleRequestContext

File: src/RequestContext.php [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * @var string
+     */
+    protected $method = null;
+
+    /**
+     * @var array
+     */
+    protected $attributes = [];
```
Comment (funivan) [inline src/RequestContext.php]: $data

File: src/RequestContext.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    /**
+     * @param string $method
+     * @param array $attributes
+     */
+    public function __construct($method, array $attributes = []) {
+      if (!is_string($method)) {
+        throw new \InvalidArgumentException('Parameter "method" should be a string,  ' . gettype($method) . ' given.');
+      }
+      $method = mb_strtoupper($method);
```
Comment (funivan) [inline src/RequestContext.php]: strtoupper. We do not need mb_ function

File: src/RequestContext.php [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      if (!in_array($method, [self::METHOD_GET, self::METHOD_POST])) {
+        throw new \InvalidArgumentException('Invalid http method name.');
+      }
+      $this->method = $method;
+      $this->attributes = $attributes;
+    }
+
+
+    /**
+     * @return mixed
```
Comment (funivan) [inline src/RequestContext.php]: return string

File: src/RequestContext.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      return isset($this->attributes[$name]);
+    }
+
+
+    /**
+     * @param string $name
+     * @return mixed|null
+     */
+    public function get($name) {
+      return isset($this->attributes[$name]) ? $this->attributes[$name] : null;
```
Comment (funivan) [inline src/RequestContext.php]: if we send <input name="bonus" value="0">    if we submit this element then get will return null but not 0

File: src/RequestContext.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+
+    /**
+     * @return array
+     */
+    public function all() {
```
Comment (funivan) [inline src/RequestContext.php]: getData

File: src/Form.php [hunk 12]
```diff
@@ -66,6 +66,26 @@ public function setName($name) {
...
+     * @param RequestContext $requestContext
+     * @return $this
+     */
+    public function handleRequestContext(RequestContext $requestContext) {
+      $this->cleanValidationFlag();
+      foreach ($this->getElements() as $element) {
+        $element->handleRequestContext($requestContext);
+      }
+
+      $this->data = $requestContext->all();
```
Comment (funivan) [inline src/Form.php]: If form contain 1 field `<input name="name" >` But user send 20 fields with post. Then this-data contains all post data. I think it should contains only 1 field `name`

File: src/Form.php [hunk 13]
```diff
@@ -81,6 +101,7 @@ public function getData() {
     /**
      * @param array|\Iterator $data
      * @throws \Exception
+     * @deprecated
      */
     public function setData($data) {
```
Comment (funivan) [inline src/Form.php]: add trigger_error

File: CHANGELOG.md (hunk 4 above)
Comment (sevikerr) [inline CHANGELOG.md]: CheckBoxList::getValue return all selected values. But what if you want to know some value is checked?

File: src/Form.php (hunk 13 above)
Comment (sevikerr) [inline src/Form.php]: fixed

**Pull request merged**